### PR TITLE
[PS5][Driver] Link main components with -pie by default

### PR DESCRIPTION
**Pull request by playstation-edd**
The PS5 linker currently forces `-pie` for typical link jobs. Have the driver pass `pie` under the same conditions. With this change we can remove our private linker patch and also allow `-no-pie` to have an effect.

SIE tracker: TOOLCHAIN-16704

### CODE REVIEW

**Review by pogo59 — APPROVED** [review]
LGTM

File: clang/lib/Driver/ToolChains/PS4CPU.cpp [hunk 1]
```diff
@@ -237,7 +237,10 @@ void tools::PS5cpu::Linker::ConstructJob(Compilation &C, const JobAction &JA,
   if (!D.SysRoot.empty())
     CmdArgs.push_back(Args.MakeArgString("--sysroot=" + D.SysRoot));
 
-  if (Args.hasArg(options::OPT_pie))
+  // Default to PIE for non-static main components.
```
Comment (MaskRay) [inline clang/lib/Driver/ToolChains/PS4CPU.cpp:240]: While the generic ABI refers to executable and shared object files as "components", "main component" isn't used.

main executable or just executable

File: clang/test/Driver/ps5-linker.c [hunk 2]
```diff
@@ -1,3 +1,18 @@
+// Test that PIE is the default for main components
+
+// RUN: %clang --target=x86_64-scei-ps5 %s -### 2>&1 | FileCheck --check-prefixes=CHECK-PIE %s
+
+// CHECK-PIE: {{ld(\.exe)?}}"
+// CHECK-PIE-SAME: "-pie"
+
+// RUN: %clang --target=x86_64-scei-ps5 -no-pie %s -### 2>&1 | FileCheck --check-prefixes=CHECK-NO-PIE %s
+// RUN: %clang --target=x86_64-scei-ps5 -r %s -### 2>&1 | FileCheck --check-prefixes=CHECK-NO-PIE %s
+// RUN: %clang --target=x86_64-scei-ps5 -shared %s -### 2>&1 | FileCheck --check-prefixes=CHECK-NO-PIE %s
```
Comment (MaskRay) [inline clang/test/Driver/ps5-linker.c:10]: Suggest that the `-shared` RUN line additionally checks `-shared`.

(I'm starting a 3-week vacation this Friday and will have limited availability.)

File: clang/lib/Driver/ToolChains/PS4CPU.cpp (hunk 1 above)
Comment (playstation-edd) [reply clang/lib/Driver/ToolChains/PS4CPU.cpp]: Done!

File: clang/test/Driver/ps5-linker.c (hunk 2 above)
Comment (playstation-edd) [reply clang/test/Driver/ps5-linker.c]: Done!

**Pull request merged**